velox/functions/sparksql/SIMDJsonFunctions.h

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/prestosql/SIMDJsonFunctions.h"
+
+using namespace simdjson;
+
+namespace facebook::velox::functions::sparksql {
+
+template <typename T>
+struct SIMDGetJsonObjectFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& config,
+      const arg_type<Varchar>* /*json*/,
+      const arg_type<Varchar>* jsonPath) {
+    if (jsonPath != nullptr) {
+      formattedJsonPath_ = getFormattedJsonPath(*jsonPath);
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& json,
+      const arg_type<Varchar>& jsonPath) {
+    ParserContext ctx(json.data(), json.size());
+    ctx.parseDocument();
+    auto rawResult = formattedJsonPath_.has_value()
+        ? ctx.jsonDoc.at_pointer(formattedJsonPath_.value().data())
+        : ctx.jsonDoc.at_pointer(getFormattedJsonPath(jsonPath).data());
+    if (rawResult.error()) {
+      return false;
+    }
+    auto error = extractStringResult(rawResult, result);
+    if (error) {
+      return false;
+    }
+
+    const char* currentPos;
+    ctx.jsonDoc.current_location().get(currentPos);
+    return isValidEndingCharacter(currentPos);
+  }
+
+ private:
+  // Makes a conversion from spark's json path, e.g., converts
+  // "$.a.b" to "/a/b".
+  FOLLY_ALWAYS_INLINE std::string getFormattedJsonPath(
+      const arg_type<Varchar>& jsonPath) {
+    char formattedJsonPath[jsonPath.size() + 1];
+    int j = 0;
+    for (int i = 0; i < jsonPath.size(); i++) {
```

Comment:
This logic is pretty important as it effectively defines what paths are supported. It would be nice to extract it into a separate header with clear documentation and dedicated unit tests.

For example, looks like we allow $ to appear anywhere within the path. Is this correct? Can you provide an example of a valid 'path' where $ appears somewhere in the middle?

Reply:
>  looks like we allow $ to appear anywhere within the path. Is this correct?

Just found this code is wrong. `$` should be in the initial position and should not occur after this, which is verified with Spark query. I just updated this code and added some test.